Offline/OfflinePlayer/objc/OfflinePlayer/DownloadsViewController.m

```diff
@@ -0,0 +1,1181 @@
+    [updatedOfflineVideoTokenArray removeObject:offlineVideoToken];
+    gVideosViewController.offlineVideoTokenArray = updatedOfflineVideoTokenArray;
+    
+    [self.downloadsTableView deleteRowsAtIndexPaths:@[indexPath]
+                                   withRowAnimation:UITableViewRowAnimationFade];
+    
+    if (self.currentlyPlayingOfflineVideoToken != nil
+        && [self.currentlyPlayingOfflineVideoToken isEqualToString:offlineVideoToken])
+    {
+        // Hide this video if it was playing
+        [self doPlayHideButton:nil];
+    }
+    
+    // Remove poster image:
+    self.posterImageView.image = nil;
+
+    // Update text in info panel
+    [self updateInfoForSelectedDownload];
+    
+    [gVideosViewController updateStatus];
+    [self refresh];
+}
+
+- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
+{
+    return [NSString stringWithFormat:@"%d Offline Videos", (int)gVideosViewController.offlineVideoTokenArray.count];
+}
+
+- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section
+{
+    NSArray<BCOVOfflineVideoStatus *> *statusArray = [gVideosViewController.offlineVideoManager offlineVideoStatus];
+    
+    int inProgressCount = 0;
+    for (BCOVOfflineVideoStatus *offlineVideoStatus in statusArray)
+    {
+        if (offlineVideoStatus.downloadState == BCOVOfflineVideoDownloadStateDownloading)
+        {
+            inProgressCount ++;
+        }
+    }
+
+    NSString *footerString;
+    
+    switch (inProgressCount)
+    {
+        case 0:
+            footerString = [NSString stringWithFormat:@"All Videos Are Fully Downloaded"];
+            break;
+        case 1:
+            footerString = [NSString stringWithFormat:@"1 Video Is Still Downloading"];
+            break;
+        default:
+            footerString = [NSString stringWithFormat:@"%d Videos Are Still Downloading",
+                            inProgressCount];
+            break;
+    }
+    
+    return footerString;
+}
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    int index = (int)indexPath.row;
+    BCOVOfflineVideoToken offlineVideoToken = gVideosViewController.offlineVideoTokenArray[index];
+    BCOVOfflineVideoStatus *offlineVideoStatus = [gVideosViewController.offlineVideoManager offlineVideoStatusForToken:offlineVideoToken];
+
+    BCOVVideo *video = [BCOVOfflineVideoManager.sharedManager videoObjectFromOfflineVideoToken:offlineVideoToken];
+    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"download_cell"
+                                                            forIndexPath:indexPath];
+    cell.textLabel.text = video.properties[@"name"];
+    cell.textLabel.textColor = (video.usesFairPlay ? [UIColor colorWithRed:0.75 green:0.0 blue:0.0 alpha:1.0] : UIColor.blackColor);
```

Comment:
Add comment:
"// Use red label to indicate that the video is protected with FairPlay"

Reply:
√